### ISSUE TITLE
How to deal with categorical variables in this method

### ISSUE THREAD

**Issue by HaoXiaoQB** [jsyoon0823/GAIN#10]
Hi Jinsung,

Thanks for developing this method! It's very cool and useful. 

In supplement material, I saw you showed the imputation performance for categorical variables. How did you deal with categorical variables in this method?

I tried to transform the dataset with One Hot Encoding and send to the network without any modification, but the imputation result (false prediction rate) is very similar to random guess. 

Hao

**Comment by richardwu** [comment]
I've also tried the same thing with categorical variables and the model seems to suffer from mode collapse (within a categorical variable). Seeking some guidance on this.

**Comment by redfungus** [comment]
HI!

Thank you for providing the code to your paper. 
I've realized that in the implementation, the loss function used for the reconstruction loss only considers continuous variables in contrast with the paper that considers both cases. 
Do you happen to have the code for the categorical variables too?

**Comment by HaoXiao2018** [comment]
Don't have the code but these are the tips from the author.

On top of one-hot encoding, to make categorical variable work:
(1) Activation function: use softmax activation for each spanned categorical vector (instead of sigmoid activation)
(2) Change the MSE loss to the cross-entropy loss for the categorical variable.

**Comment by jsyoon0823** [comment]
Thanks, HaoXiao2018.
This is exactly what I did for the categorical variables.
As can be seen in the paper, I used cross-entropy for categorical variable and mse for continuous variable.
But more important thing is that we need to use separate output layer (activation function with softmax).

**Comment by redfungus** [comment]
So, is it possible to have both categorical and continuous features in a dataset?

**Comment by jsyoon0823** [comment]
Yes. It would be better if you modify the codes based on the comments from HaoXiao2018 above and apply GAIN for your mixed-type data.

**Comment by samyakag** [comment]
If I have categorical data, does that mean I also have to change the dimensions of all the network layers, etc as well?

**Comment by jsyoon0823** [comment]
Network parameters should be updated for different datasets.
Even for the datasets which only have continuous variables, those hyper-parameters should be optimized to maximize the performance.
